### PR TITLE
check and fix inconsistency between TiKV and TiDB's cast (part 1)

### PR DESCRIPTION
**Pull request by H-ZeX**
Signed-off-by: H-ZeX [hzx20112012@gmail.com](mailto:hzx20112012@gmail.com)
 
 ### What have you changed?

compare tidb and tikv's impl of `cast` and fix what they are inconsistent.

 ### What is the type of the changes?

 * Improvement (a change which is an improvement to an existing feature)
 * Bugfix (a change which fixes an issue)
 * Engineering (engineering change which doesn't change any feature or fix any issue)
 
 ### How is the PR tested?

uint test
 
 ### Does this PR affect documentation (docs) or should it be mentioned in the release notes?
 
No.
 
 ### Does this PR affect `tidb-ansible`?

 No.

### CODE REVIEW

File: components/tidb_query/src/codec/datum.rs [hunk 1]
```diff
@@ -268,9 +271,9 @@ impl Datum {
             Datum::I64(i) => Some(i != 0),
             Datum::U64(u) => Some(u != 0),
             Datum::F64(f) => Some(f.round() != 0f64),
-            Datum::Bytes(ref bs) => {
-                Some(!bs.is_empty() && bs.to_int(ctx, FieldTypeTp::LongLong)? != 0)
-            }
+            Datum::Bytes(ref bs) => Some(
+                !bs.is_empty() && <&[u8] as ConvertTo<i64>>::convert(&bs.as_slice(), ctx)? != 0,
```
Comment (breeswish) [inline components/tidb_query/src/codec/datum.rs]: If this is changed, what about these things in `as_mysql_bool`?

Comment (H-ZeX) [reply components/tidb_query/src/codec/datum.rs]: Actually, this convert call `to_int(ctx, FieldTypeTp::LongLong)`, so their behavior is same.

I change `to_int(ctx, FieldTypeTp::LongLong)` to `convert` is to avoid `FieldTypeTp:LongLong` which is confused:  why we use `FieldTypeTp::LongLong` but not other tp, we can hide this with     using `convert`

File: components/tidb_query/src/codec/datum.rs [hunk 2]
```diff
@@ -192,7 +192,7 @@ impl Datum {
                 Ok(d.cmp(&d2))
             }
             _ => {
-                let f = bs.convert(ctx)?;
+                let f: f64 = bs.convert(ctx)?;
```
Comment (Deardrops) [inline components/tidb_query/src/codec/datum.rs]: Do we need declare type `f64` here? If need, please add `f64` to every case like this.

File: components/tidb_query/src/codec/mysql/decimal.rs [hunk 3]
```diff
@@ -1793,20 +1794,13 @@ impl ConvertTo<Decimal> for Real {
 }
 
 impl ConvertTo<Decimal> for &[u8] {
+    // FIXME, the err handle is not exactly same as TiDB's,
+    //  TiDB's seems has bug, fix this after fix TiDB's
```
Comment (Deardrops) [inline components/tidb_query/src/codec/mysql/decimal.rs]:
```suggestion
    // TiDB's seems has bug, fix this after fix TiDB's
```

File: components/tidb_query/src/codec/mysql/duration.rs [hunk 4]
```diff
@@ -1126,6 +1135,183 @@ mod tests {
...
+            ),
+            // will truncated
+            (8376049, 0, Err(Error::truncated_wrong_val("", "")), false),
+            (8375960, 0, Err(Error::truncated_wrong_val("", "")), false),
+            (8376049, 0, Err(Error::truncated_wrong_val("", "")), false),
+            // TODO, add test for num>=10000000000
+            //  after Duration::from_f64 had impl logic for num>=10000000000
+            // (10000000000, 0, Ok(Duration::parse("0:0:0".as_bytes(), 0).unwrap())),
+            // (10000235959, 0, Ok(Duration::parse("23:59:59".as_bytes(), 0).unwrap())),
+            // (10000000000, 0, Ok(Duration::parse("0:0:0".as_bytes(), 0).unwrap())),
```
Comment (Deardrops) [inline components/tidb_query/src/codec/mysql/duration.rs]: I suggest add test case for num>=10000000000 here (assert the wrong behavior in TiKV).
In case, If someone implement this funciton or modify the logic, the test case will failed, and it will inform he/she to remove the old test case and use new test case.

File: components/tidb_query/src/codec/convert.rs [hunk 5]
```diff
@@ -151,64 +151,67 @@ pub fn truncate_binary(s: &mut Vec<u8>, flen: isize) {
...
+    let shift = 10f64.powi(i32::from(decimal));
+    let max_f = 10f64.powi(i32::from(flen - decimal)) - 1.0 / shift;
+
+    if !f.is_infinite() {
         let tmp = f * shift;
         if tmp.is_finite() {
             f = tmp.round() / shift
         }
-    };
+    }
```
[the remark this answers is not in the document]
Comment (H-ZeX) [reply components/tidb_query/src/codec/convert.rs]: I think it is unnecessary.

File: components/tidb_query/src/codec/datum.rs (hunk 2 above)
Comment (H-ZeX) [reply components/tidb_query/src/codec/datum.rs]: Here we need, if in somewhere, it is necessary but we forget to add, rustc will report an error.

File: components/tidb_query/src/codec/mysql/decimal.rs (hunk 3 above)
Comment (H-ZeX) [reply components/tidb_query/src/codec/mysql/decimal.rs]: Add this space to indicate this is belong to the section of `FIXME`, and in CLion, it will look like this
image

File: components/tidb_query/src/codec/mysql/duration.rs [hunk 6]
```diff
@@ -695,7 +680,32 @@ impl Duration {
...
+        (Some(dur), None)
+    }
+
+    pub fn from_i64(ctx: &mut EvalContext, n: i64, fsp: u8) -> Result<Duration> {
+        let (dur, err) = Duration::from_i64_without_ctx(n, fsp);
+        match err {
+            Some(e) => {
+                if e.is_overflow() {
+                    ctx.handle_overflow_err(e)?;
+                    if dur.is_none() {
```
Comment (breeswish) [inline components/tidb_query/src/codec/mysql/duration.rs]: Then you should use unwrap?

File: components/tidb_query/src/codec/mysql/duration.rs [hunk 7]
```diff
@@ -708,20 +718,16 @@ impl ConvertTo<f64> for Duration {
 }
 
 impl ConvertTo<Decimal> for Duration {
+    /// This function should not return err,
+    /// if it return err, then the err is because of bug.
```
Comment (breeswish) [inline components/tidb_query/src/codec/mysql/duration.rs]: ditto. Unwrap?

File: components/tidb_query/src/codec/mysql/decimal.rs [hunk 8]
```diff
@@ -1710,21 +1720,17 @@ enable_conv_for_int!(usize, u64);
 enable_conv_for_int!(isize, i64);
 
 impl ConvertTo<f64> for Decimal {
+    /// This function should not return err,
+    /// if it return err, then the err because of bug.
+    ///
+    /// Port from TiDB's MyDecimal::ToFloat64.
```
[the remark this answers is not in the document]
Comment (H-ZeX) [reply components/tidb_query/src/codec/mysql/decimal.rs]: However, the parse function is very likely to has some small bug, but such bug is not fatal, so if we `unwrap` here, the tikv will panic because of such small bug, which is very bad for our user

[the remark this answers is not in the document]
Comment (H-ZeX) [reply components/tidb_query/src/codec/mysql/decimal.rs]: Should we construct a mechanism to handle such error which because of some bug, or we will loss  info of such some bug when meet it.

File: components/tidb_query/src/codec/mysql/decimal.rs [hunk 9]
```diff
@@ -1547,6 +1547,16 @@ impl Decimal {
         Res::Ok(x)
     }
 
+    // TODO, add test
```
Comment (mapleFU) [inline components/tidb_query/src/codec/mysql/decimal.rs]: So, corresponding tests are in other pr?

Comment (H-ZeX) [reply components/tidb_query/src/codec/mysql/decimal.rs]: No, I will add test for it latter in this PR.

File: components/tidb_query/src/codec/convert.rs [hunk 10]
```diff
@@ -304,14 +335,25 @@ impl ToInt for &[u8] {
...
+        if s.starts_with('-') {
+            ctx.handle_overflow_err(Error::overflow("BIGINT UNSIGNED", &s))?;
+            return Ok(0);
+        }
+        let val = s.parse::<u64>();
         match val {
             Ok(val) => val.to_uint(ctx, tp),
             Err(_) => {
-                ctx.handle_overflow_err(Error::overflow("BIGINT UNSIGNED", &vs))?;
+                ctx.handle_overflow_err(Error::overflow("BIGINT UNSIGNED", &s))?;
```
Comment (iosmanthus) [inline components/tidb_query/src/codec/convert.rs]: `Error::overflow("BIGINT UNSIGNED", s)` is enough.

File: components/tidb_query/src/codec/convert.rs [hunk 11]
```diff
@@ -1920,8 +2028,9 @@ mod tests {
         let cfg = EvalConfig::from_flag(Flag::TRUNCATE_AS_WARNING);
         let mut ctx = EvalContext::new(Arc::new(cfg));
         let mut ft = FieldType::default();
-        <FieldType as FieldTypeAccessor>::set_tp(&mut ft, FieldTypeTp::String);
-        <FieldType as FieldTypeAccessor>::set_collation(&mut ft, Collation::Binary);
+        let fta = ft.as_mut_accessor();
```
Comment (iosmanthus) [inline components/tidb_query/src/codec/convert.rs]: why not just produce a chain call?
```rust
ft.as_mut_accessor().set_tp(...).set_collation(...);
```

File: components/tidb_query/src/codec/mysql/decimal.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (iosmanthus) [reply components/tidb_query/src/codec/mysql/decimal.rs]: Please consider removing this comment. If it fails to parse, no matter what reason `tikv` should not fail because of some invalid data format. We need to return an error.

File: components/tidb_query/src/codec/mysql/decimal.rs [hunk 12]
```diff
@@ -1793,20 +1794,13 @@ impl ConvertTo<Decimal> for Real {
...
-                ctx.handle_overflow_err(Error::overflow("DECIMAL", ""))?;
-                d
-            }
-            Res::Truncated(d) => {
-                ctx.handle_truncate(true)?;
-                d
-            }
-        };
-        Ok(dec)
+        let r = Decimal::from_bytes(self)?;
```
Comment (iosmanthus) [inline components/tidb_query/src/codec/mysql/decimal.rs]: One-row task split in three steps XD

File: components/tidb_query/src/codec/mysql/decimal.rs [hunk 13]
```diff
@@ -1824,6 +1818,32 @@ impl ConvertTo<Decimal> for Bytes {
     }
 }
 
+impl ConvertTo<Decimal> for Json {
+    /// Port from TiDB's types.ConvertJSONToDecimal
+    #[inline]
+    fn convert(&self, ctx: &mut EvalContext) -> Result<Decimal> {
+        match self {
+            Json::String(s) => {
+                match Decimal::from_str(s.as_str()) {
```
Comment (iosmanthus) [inline components/tidb_query/src/codec/mysql/decimal.rs]: Consider use `or_else`. Please don't repeat yourself. Those combinators make your code more concise.

File: components/tidb_query/src/codec/mysql/duration.rs [hunk 14]
```diff
@@ -695,7 +680,32 @@ impl Duration {
...
+    pub fn from_i64(ctx: &mut EvalContext, n: i64, fsp: u8) -> Result<Duration> {
+        let (dur, err) = Duration::from_i64_without_ctx(n, fsp);
+        match err {
+            Some(e) => {
+                if e.is_overflow() {
+                    ctx.handle_overflow_err(e)?;
+                    if let Some(d) = dur {
+                        Ok(d)
+                    } else {
+                        Err(box_err!("Expect a not none result here, this is a bug"))
```
Comment (mapleFU) [inline components/tidb_query/src/codec/mysql/duration.rs]: Should we use `unreachable!` here?

Comment (H-ZeX) [reply components/tidb_query/src/codec/mysql/duration.rs]: https://github.com/tikv/tikv/pull/5423#discussion_r323102836

> However, the parse function is very likely to has some small bug, but such bug is not fatal, so if we unwrap here, the tikv will panic because of such small bug, which is very bad for our user

File: components/tidb_query/src/codec/mysql/decimal.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (H-ZeX) [reply components/tidb_query/src/codec/mysql/decimal.rs]:
```rust
impl ConvertTo<f64> for Decimal {
    /// This function should not return err,
    /// if it return err, then the err because of bug.
    ///
    /// Port from TiDB's MyDecimal::ToFloat64.
    #[inline]
    fn convert(&self, _: &mut EvalContext) -> Result<f64> {
        let val = self.to_string().parse()?;
        Ok(val)
    }
}
```

the string is from `decimal::to_string`, so it should not be invalid string

File: components/tidb_query/src/codec/mysql/decimal.rs (hunk 12 above)
Comment (H-ZeX) [reply components/tidb_query/src/codec/mysql/decimal.rs]: without split, after rustfmt, it will look ungly
image

File: components/tidb_query/src/codec/mysql/decimal.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (breeswish) [reply components/tidb_query/src/codec/mysql/decimal.rs]: I think for any "should" they deserve an assertion. You shall either use assertion for your "should", or change comment to "may".

File: components/tidb_query/src/codec/mysql/duration.rs [hunk 15]
```diff
@@ -646,46 +647,30 @@ impl Duration {
...
-
+    /// If the error is overflow, the result will return true,
+    /// otherwise, only one of result or err will be returned
+    pub fn from_i64_without_ctx(mut n: i64, fsp: u8) -> (Option<Duration>, Option<Error>) {
         if n > i64::from(MAX_DURATION_VALUE) || n < -i64::from(MAX_DURATION_VALUE) {
             // FIXME: parse as `DateTime` if `n >= 10000000000`
-            ctx.handle_overflow_err(Error::overflow("Duration", &n.to_string()))?;
             let max = Duration::new(n < 0, MAX_HOURS, MAX_MINUTES, MAX_SECONDS, 0, fsp);
-            return Ok(max);
+            return (Some(max), Some(Error::overflow("Duration", &n.to_string())));
```
Comment (iosmanthus) [inline components/tidb_query/src/codec/mysql/duration.rs]: Ditto

File: components/tidb_query/src/codec/mysql/duration.rs [hunk 16]
```diff
@@ -695,7 +680,32 @@ impl Duration {
...
             fsp,
         );
-        Ok(dur)
+        (Some(dur), None)
+    }
+
+    pub fn from_i64(ctx: &mut EvalContext, n: i64, fsp: u8) -> Result<Duration> {
+        let (dur, err) = Duration::from_i64_without_ctx(n, fsp);
+        match err {
+            Some(e) => {
```
Comment (iosmanthus) [inline components/tidb_query/src/codec/mysql/duration.rs]: Please don't repeat yourself

File: components/tidb_query/src/codec/mysql/decimal.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (H-ZeX) [reply components/tidb_query/src/codec/mysql/decimal.rs]: Ok, I will add `debug_assert!` in such place, and I think the doc is necessary

File: components/tidb_query/src/codec/datum.rs [hunk 17]
```diff
@@ -207,6 +207,7 @@ impl Datum {
                 Ok(d.cmp(dec))
             }
             _ => {
+                // FIXME, here is not same as TiDB's
```
Comment (lonng) [inline components/tidb_query/src/codec/datum.rs]: It's better to keep a consistent style of `FIXME: xxx` with other snippets in this project.

Comment (lonng) [reply components/tidb_query/src/codec/datum.rs]: And please clarify the difference between TiDB and TiKV. (Otherwise it's a useless comment because it cannot help the reviewer to review)

File: components/tidb_query/src/codec/convert.rs [hunk 18]
```diff
@@ -151,64 +151,66 @@ pub fn truncate_binary(s: &mut Vec<u8>, flen: isize) {
 /// `truncate_f64` (`TruncateFloat` in TiDB) tries to truncate f.
 /// If the result exceeds the max/min float that flen/decimal
 /// allowed, returns the max/min float allowed.
+///
+/// This function can only return `Res::Ok` or `Res::Overflow`, no `Res::Truncated` will be returned.
```
Comment (lonng) [inline components/tidb_query/src/codec/convert.rs]: It's better to revert this change.

File: components/tidb_query/src/codec/convert.rs [hunk 19]
```diff
@@ -228,47 +230,72 @@ impl ToInt for u64 {
...
             }
         }
-
         let upper_bound = integer_unsigned_upper_bound(tp);
         if val > upper_bound as f64 {
-            ctx.handle_overflow_err(overflow!(val, upper_bound))?;
-            return Ok(upper_bound);
+            ctx.handle_overflow_err(overflow(val, tp))?;
+            Ok(upper_bound)
+        } else if val == upper_bound as f64 {
```
Comment (lonng) [inline components/tidb_query/src/codec/convert.rs]: Seems we can simplify this code:
```rust
if val > upper_bound as f64 {
    ctx.handle_overflow_err(overflow(val, tp))?;
    return Ok(upper_bound);	            Ok(upper_bound)
} else {
    Ok(val as u64)
}
```
See: https://gist.github.com/4c3bcdc9ec36c60bd5920f6be138918b

File: components/tidb_query/src/codec/convert.rs [hunk 20]
```diff
@@ -304,14 +335,25 @@ impl ToInt for &[u8] {
...
     fn to_uint(&self, ctx: &mut EvalContext, tp: FieldTypeTp) -> Result<u64> {
         let s = get_valid_utf8_prefix(ctx, self)?;
-        let vs = get_valid_int_prefix(ctx, s)?;
-        let val = vs.parse::<u64>();
+        let s = s.trim();
+        let s = get_valid_int_prefix(ctx, s)?;
+        // in TiDB, it use strconv.ParseUint here,
+        // strconv.ParseUint will return 0 and a err if the str is neg
+        if s.starts_with('-') {
+            ctx.handle_overflow_err(Error::overflow("BIGINT UNSIGNED", &s))?;
```
Comment (lonng) [inline components/tidb_query/src/codec/convert.rs]: Why use `&s`?

File: components/tidb_query/src/codec/convert.rs [hunk 21]
```diff
@@ -344,26 +386,43 @@ impl ToInt for Decimal {
...
+        if tp == FieldTypeTp::LongLong {
+            Ok(r)
+        } else {
+            r.to_uint(ctx, tp)
+        }
     }
 }
 
 impl ToInt for DateTime {
+    // FiXME
```
Comment (lonng) [inline components/tidb_query/src/codec/convert.rs]: ditto. (unify the `FIXME: xxx` style)

File: components/tidb_query/src/codec/mysql/decimal.rs [hunk 22]
```diff
@@ -1793,20 +1794,13 @@ impl ConvertTo<Decimal> for Real {
...
-                d
-            }
-            Res::Truncated(d) => {
-                ctx.handle_truncate(true)?;
-                d
-            }
-        };
-        Ok(dec)
+        let r = Decimal::from_bytes(self)?;
+        let err = Error::overflow("DECIMAL", "");
```
Comment (lonng) [inline components/tidb_query/src/codec/mysql/decimal.rs]: Is `Error::overflow("DECIMAL", self);` better?

File: components/tidb_query/src/codec/mysql/duration.rs [hunk 23]
```diff
@@ -646,46 +647,30 @@ impl Duration {
...
-        buf
+        self.format("")
     }
 
-    pub fn from_i64(ctx: &mut EvalContext, mut n: i64, fsp: u8) -> Result<Duration> {
-        use crate::codec::error::ERR_TRUNCATE_WRONG_VALUE;
-
+    /// If the error is overflow, the result will be returned, too.
+    /// Otherwise, only one of result or err will be returned
+    pub fn from_i64_without_ctx(mut n: i64, fsp: u8) -> (Option<Duration>, Option<Error>) {
```
Comment (lonng) [inline components/tidb_query/src/codec/mysql/duration.rs]: Seems this function can be a private method.

File: components/tidb_query/src/expr/builtin_cast.rs [hunk 24]
```diff
@@ -147,68 +146,98 @@ impl ScalarFunc {
...
+        let val: Cow<Json> = try_opt!(self.children[0].eval_json(ctx, row));
         let res = val.to_int(ctx, FieldTypeTp::LongLong)?;
         Ok(Some(res))
     }
 
     pub fn cast_int_as_real(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<f64>> {
         let val = try_opt!(self.children[0].eval_int(ctx, row));
         if !self.children[0].is_unsigned() {
-            Ok(Some(self.produce_float_with_specified_tp(ctx, val as f64)?))
+            Ok(Some(produce_float_with_specified_tp(
```
Comment (lonng) [inline components/tidb_query/src/expr/builtin_cast.rs]: Prefer:
```rust
let val = if !self.children[0].is_unsigned() {
    val as f64
} else {
    val as u64 as f64
};
Ok(Some(produce_float_with_specified_tp(
    ctx,
    &self.field_type,
    val,
 )?))
```

File: components/tidb_query/src/codec/datum.rs (hunk 17 above)
Comment (H-ZeX) [reply components/tidb_query/src/codec/datum.rs]: There is many difference here, so I place a FIXME here. Compare it with `Datum::compareMysqlDecimal` for detail difference

File: components/tidb_query/src/codec/mysql/duration.rs (hunk 23 above)
Comment (H-ZeX) [reply components/tidb_query/src/codec/mysql/duration.rs]: I split out this func from `from_i64` is for the `cast_int_as_duration`, so I should be `pub`

File: components/tidb_query/src/codec/mysql/decimal.rs (hunk 22 above)
Comment (H-ZeX) [reply components/tidb_query/src/codec/mysql/decimal.rs]: &[u8] has not imp Display trait, so we should construct a string, however, the cost of construcut a string is not small
